factories/docFactory.js

```diff
@@ -8,55 +8,65 @@
 
   function DocFactory() {
     var Doc = function(doc, opts) {
-      var self        = this;
-
+      var self = this;
       angular.copy(doc, self);
+      self.doc = doc;
+      self.opts = opts;
+    };
 
-      self.doc             = doc;
-
-      self.groupedBy       = groupedBy;
-      self.group           = group;
-      self.options         = options;
-      self.version         = version;
-      self.fieldsAttrName  = fieldsAttrName;
-      self.fieldsProperty  = fieldsProperty;
+    Doc.prototype = {};
```

Comment:
Can you explain why this change w as made?  Is htis a "better" way to  do this?   Are there other places through the codebase that should be like this?

Reply:
This change was made so that the methods can be overridden in subclasses. This is a technique called "prototype inheritance", and yes, it should be used everywhere on all "classes" in JavaScript. I believe there are other places in the code base where we already store the methods on the prototype (instead of creating new methods on each instantiated object). However, if you want to "modernize" this, you should instead migrate the code to [ES6 classes](https://caniuse.com/es6-class).

Reply:
Thanks for sharing this!   I think I'd like to seperate what you've provided into two PR's if I can..  the first is for the support for scripted fields, and the second for resolving this "prototype inheritance" issue..